python/paddle/fluid/io.py

```diff
@@ -643,6 +643,9 @@ def save_inference_model(dirname,
 
     if main_program is None:
         main_program = default_main_program()
+        if main_program.is_optimized:
+            raise RuntimeError(
+                "save_inference_model must put before you call memory_optimize.")
```

Comment:
Better to add more description for users, such as

"save_inference_model must put before you call memory_optimize."
 "the memory_optimize will modify the original program, and that is not suitable for inference"